toptal/picasso
# feat: avoid page scrollbar jump

[FX-4235]

This pull request fixes jumping width behaviour of the page. The problem is caused by scrollbar that appears if the content is high enough.

https://github.com/toptal/picasso/assets/1390758/cd6f0f78-84f7-4d21-9bc4-9c8a670cb530

The problem was already addressed in Client Portal in https://github.com/toptal/client-portal/pull/6048/files#diff-2544365cf05041a8ae61f573f230da494c339ec97e660e49f175b1684011b0baR36 (Slack discussion https://toptal-core.slack.com/archives/CEW0UBG5B/p1678095730351189), but the fix was removed after Client Portal fully moved to Picasso layout components.

### Technical details

**Why workaround with `vw`?**

Reasons
- it is approved by Design Team
- it has been already used in Client Portal

Alternatives
- have a reserved space for scrollbar so the page does not change its width (via `scrollbar-gutter` or `overflow` CSS properties) – ⚠️ rejected on design level, more in https://toptal-core.atlassian.net/browse/FX-4235?focusedCommentId=286579
- do nothing – ⚠️ rejected on design level, more in https://toptal-core.atlassian.net/browse/FX-4235?focusedCommentId=286579

**How many users are affected?**

- [Desktop Operating System Market Share Worldwide](https://gs.statcounter.com/os-market-share/desktop/worldwide/)  – 80% of desktops have this problem (mostly Windows users, MacOS does not have this problem if default scrollbar settings are enabled)
- [Mobile Operating System Market Share Worldwide](https://gs.statcounter.com/os-market-share/mobile/worldwide)  – not a problem on mobile devices, overlay scrollbars are used most of the time (at least, on iOS and Android)

### How to test


- [Temploy](https://picasso.toptal.net/fx-4235-page-header-is-jumping)
- Check out of the code sandbox at https://codesandbox.io/s/picasso-scrollbar-jump-tl4n9l?file=/src/App.tsx
- Try different page state combinations at [Client Portal temploy](https://client-portal-pr-7165.toptal.rocks/) (log in as https://staff-portal.toptal.net/clients/767902 at https://staff-portal.toptal.net/ first)
  - Open short page like https://client-portal-pr-7165.toptal.rocks/jobs
  - Open long page like https://client-portal-pr-7165.toptal.rocks/ or https://client-portal-pr-7165.toptal.rocks/team
  - Open and close modal at long page, for example, by clicking `+ Add New User` at https://client-portal-pr-7165.toptal.rocks/interviews/VjEtRW5nYWdlbWVudC0zNzQ4NjU (demo below)

<details>
  <summary>Clicking `+ Add New User` and opening modal in Client Portal</summary>

https://github.com/toptal/picasso/assets/1390758/f4a12522-ec68-49ea-a76d-3ad2f82f0cbe

</details>

### Screenshots

With fix applied (`<Picasso />`)

https://github.com/toptal/picasso/assets/1390758/7be5b246-6287-479f-80a5-89d850604e67

Without fix applied (`<Picasso preventPageWidthChangeOnScrollbar={false} />`)

https://github.com/toptal/picasso/assets/1390758/27005362-061e-4b39-be44-ee4f620f5e40

### Development checks

- [N/A] Create `examples` for component
- [N/A] Covered with tests ([visual tests](https://github.com/toptal/picasso/blob/master/docs/contribution/visual-testing.md) included)

**Breaking change**

- [N/A] codemod is created and showcased in the changeset
- [x] test alpha package of Picasso in Client Portal https://github.com/toptal/client-portal/pull/7165

> All **_development checks_** should be done and set checked to pass the
> **GitHub Bot: TODOLess** action

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run package:alpha-release` - Release alpha version
- `@toptal-anvil ping reviewers` - Ping FX team for review

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and

1. You have no extra requests.
2. You have optional requests.
   1. Add `nit:` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because

1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:

1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>


[FX-4235]: https://toptal-core.atlassian.net/browse/FX-4235?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
